kratos/includes/define.h

```diff
@@ -698,41 +698,59 @@ catch(...) { Block KRATOS_THROW_ERROR(std::runtime_error, "Unknown error", MoreI
 #endif
 #define KRATOS_REGISTER_GEOMETRY(name, reference) \
     KratosComponents<Geometry<Node>>::Add(name, reference); \
+    if(!Registry::HasItem("geometries."+Registry::GetCurrentSource()+"."+name)){                                              \
```

Comment:
Agree, but this is currently how components are added ( as a matter of fact, you can have several components from the same app or different applications with the same name as long as the types are different ):

```c++
    static void Add(const std::string& rName, const TComponentType& rComponent)
    {
        // Check if a different object was already registered with this name, since this is undefined behavior
        auto it_comp =  msComponents.find(rName);
        KRATOS_ERROR_IF(it_comp != msComponents.end() && typeid(*(it_comp->second)) != typeid(rComponent)) << "An object of different type was already registered with name \"" << rName << "\"!" << std::endl;
        msComponents.insert(ValueType(rName , &rComponent));
    }
```

Nevertheless it is true that this could cause a registered component to be unloaded by an incorrect app. Just to confirm that we can to avoid this case:

```text
App1 Loads A
App2 Loads A
App2 Unloads A <-- Problem.
```

I will change it to keep track of this case

Reply:
We are in the same page!

Reply:
Maybe I am misunderstanding something, but
```
App1 Loads A
App2 Loads A
App1 Unloads A <-- Problem, too (if you intend to keep using App2)
```
I think that what you need here is something that behaves like a counted pointer, not keeping track of who registered what.

Reply:
No, because the implementation of `KratosComponents` is misleading. 

If two applications want to add a component with the same name, provided that all checks pass, this will be executed:

```c++
msComponents.insert(ValueType(rName , &rComponent));
```

Which will insert the component if not present, but will do nothing the second time (`msComponents` is a map and `insert` only does something if the key is not existent). That implies that the moment we unload the application that has added the component, the component becomes invalid: 1 - because the reference is no longer correct, 2 - because the second app did nothing.

Reply:
Ok, but in this situation the second app is left in an invalid state. Essentially, once the first app is unloaded, anything else that shared a name with it is left in an incomplete state and cannot be trusted to work.

If this is the case, wouldn't it be simpler to just have an "unload everything" (or maybe "unload all applications, leaving core untouched") mechanism instead of keeping track of individual apps?

Reply:
Well, the second app was never in a valid state, it was referencing something it does not own (our eternal discussion if a variable can be registered by two apps). I think we don't have any example in which only a single application is unloaded.

The implementation we wanted to replace (can be found isolated in #11511) actually does what you suggest (the kernel is in charge of clearing all the components) but @RiccardoRossi commented it had a very important flaw: It cannot deal with custom types. In fact #11506 would fail if we try to register a preconditioned for example, as it is a component unknown by the kernel.

> or maybe "unload all applications, leaving core untouched"

This cannot be done (and basically is the root of why I want to use tests to change the current behavior) because creating an instance of the kernel automatically tries to register the "KratosMultiphysics" app, and while the kernel class has a lifetime with initialization and destruction, the components are static and do not. Consider this code:

```C++
void do_something() {
    auto kernel_instance = std::shared_ptr<Kratos::kernel>())
}

for (int i = 0; i < 2; i++) {
     do_something();
}
```

At `i=0` the first instance of the kernel is created, the we call:

```c++
if (!IsImported("KratosMultiphysics")) {
    this->ImportApplication(mpKratosCoreApplication);
}
```

At `i=1` If code is left untouched (no deregisters, etc...) this will work, because even if the kernel was destroyed, the list applications loaded and the list of components was never cleared, which implies that applications, and components are "eternal", once registered there is no way back. In other worlds, if the kernel was a singleton it could never be destroyed even if you wanted to. 

Now consider now what happens with the new tests (for example, but a custom workflow from the GeoMechanics or anything that does not use Python as an entry point has the same problem)

The base test code defines the life cycle that will be executed per every test:
```C++
class KRATOS_API(KRATOS_TEST_UTILS) KratosCoreFastSuite : public ::testing::Test 
{
    public:
        void SetUp() override;
        void TearDown() override;

    protected:
        KratosCoreFastSuite(): mKernel() {  // we are initializing the kernel here for every test
        }

        ~KratosCoreFastSuite() {}
        }

    private:
        Kratos::Kernel mKernel;
};
```

Now, for every app I will be able to derive this using:

```C++
// Header
class KratosStructuralMechanicsIntegrationSuite : public KratosCoreFastSuite
{
public:
    KratosStructuralMechanicsIntegrationSuite();

private:
    KratosStructuralMechanicsApplication::Pointer mpStructuralApp;
    KratosLinearSolversApplication::Pointer mpLinearSolversApp;
};

// Source 
KratosStructuralMechanicsIntegrationSuite::KratosStructuralMechanicsIntegrationSuite() : KratosCoreFastSuite()
{
    mpStructuralApp = std::make_shared<KratosStructuralMechanicsApplication>();
    this->ImportApplicationIntoKernel(mpStructuralApp);
    mpLinearSolversApp = std::make_shared<KratosLinearSolversApplication>();
    this->ImportApplicationIntoKernel(mpLinearSolversApp);
}
```

Of course, I can
1 - Check if the application is already loaded and do nothing: Works without changing the current code. We use the same kernel across all tests, which is deeply wrong
2 - I unloaded the applications manually keeping the kernel: I have no way to instruct the applications to de-register custom component types and I don't know what was loaded by the kernel or by the apps
3 - I unloaded everything throw the kernel: Works, but as 2 I have no way to de-register custom components.
4 - Everyone keeps track of who is an what has registered and is in charge or de-register: No problems.

Honestly, any solution can work with we assume the consequences but for me 4 is by far the good solution. 

That being said, the fact that two applications can register the same thing and the second does nothing without a warning is super dangerous and should be changed, but probably in a separate PR.

Reply:
Ok, understood. I just think that we need to be clear that this solves the testing use-case by delegating the cleaning-up of the components to the applications, but it does not make applications unloadable in any meaningful way. Or, in other words, I do not think this solution would work for any use case that does not use python as an entrypoint, but I agree that it works for this one.

Reply:
> I just think that we need to be clear that this solves the testing use-case by delegating the cleaning-up of the components to the applications, but it does not make applications unloadable in any meaningful way

Completely agree :+1: